packages/core/src/validate/run.ts

```diff
@@ -608,6 +609,33 @@ function* getInternalFunctionStorageErrors(
   }
 }
 
+/**
+ * Gets variables declared directly in a contract and its struct definitions, or in a struct definition.
+ */
+function getVariableDeclarations(
+  contractOrStructDef: ContractDefinition | StructDefinition,
+  visitedNodeIds: Set<number>,
+): VariableDeclaration[] {
+  const results: VariableDeclaration[] = [];
+  if (contractOrStructDef.nodeType === 'ContractDefinition') {
+    for (const node of contractOrStructDef.nodes) {
+      if (node.nodeType === 'VariableDeclaration') {
+        results.push(node);
+      } else if (node.nodeType === 'StructDefinition' && !visitedNodeIds.has(node.id)) {
+        visitedNodeIds.add(node.id);
+        results.push(...getVariableDeclarations(node, visitedNodeIds));
+      }
```

Comment:
Is this necessary? It should be handled automatically by `getInternalFunctionStorageErrors`. Structs are not necessarily in storage, and those that are via EIP-7201 are already handled differently.

Reply:
I found this was necessary, otherwise this test case
```
contract StructInternalFunctionPointerUsed {
    struct S {
        function(bool) internal foo;
    }
    S s;
}
```
would give two identical errors: once for the struct def, and once for the struct variable referencing the struct def.

Reply:
Right, I think this is consistent with what I'm saying. I would remove the lines that I commented on here. As far as I can tell that would only visit the struct via dereferencing the type of `s`.

Reply:
My above comment was for the need to track `visitedNodeIds`.  I now see what you mean by removing these lines, but I don't think that would work.

`getInternalFunctionStorageErrors` first calls `getVariableDeclarations` on a contract.
- For contracts with EIP-7201 structs, the lines that you commented on are needed.  If we removed these lines, the struct definitions would not be accessed (since there are no references)
- For contracts with variables that reference regular structs, this returns those variables and `getInternalFunctionStorageErrors` will handle dereferencing those structs.

Perhaps the code layout is somewhat confusing with the multiple stages of generators (`getVariableDeclarations` and `getInternalFunctionStorageErrors`) and the overloading of their argument `contractOrStructDef: ContractDefinition | StructDefinition`

Reply:
Oh or is that only for layout purposes? I see.